sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -2493,18 +2480,30 @@ bool doesDevSupportDeviceRequirements(const device &Dev,
         DeviceBinaryProperty(*(ReqdWGSizePropIt.value())).asByteArray();
     // Drop 8 bytes describing the size of the byte array.
     ReqdWGSize.dropBytes(8);
-    int ReqdWGSizeAllDimsTotal = 1;
-    std::vector<int> ReqdWGSizeVec;
+    uint64_t ReqdWGSizeAllDimsTotal = 1;
+    std::vector<uint64_t> ReqdWGSizeVec;
     int Dims = 0;
     while (!ReqdWGSize.empty()) {
-      int SingleDimSize = ReqdWGSize.consume<int>();
-      ReqdWGSizeAllDimsTotal *= SingleDimSize;
+      // The reqd_work_group_size data is stored as uint32_t's,
+      // but we'll widen the result to uint64_t.
+      uint64_t SingleDimSize = ReqdWGSize.consume<uint32_t>();
+      if (auto res = multiply_with_overflow_check(ReqdWGSizeAllDimsTotal,
+                                                  SingleDimSize))
+        ReqdWGSizeAllDimsTotal = *res;
+      else
+        return sycl::exception(sycl::errc::kernel_not_supported,
+                               "Required work-group size is not supported"
+                               " (too large)");
```

Comment:
I wonder if this diagnostic should also be added to FE, because that is something we can check at compile-time.

I would suggest replacing "too large" with "total number of work-items requested can't fit into size_t"